aws-api/src/test/java/com/amplifyframework/api/aws/ApiAuthProvidersTest.java

```diff
@@ -0,0 +1,210 @@
+    private static final ApiConfiguration BASIC_API_CONFIG = ApiConfiguration.builder()
+                                                                             .endpoint("https://test.endpoint.com")
+                                                                             .region("us-east-1")
+                                                                             .authorizationType(
+                                                                                AuthorizationType.API_KEY
+                                                                             )
+                                                                             .build();
+    private static final CognitoUserPoolsAuthProvider DUMMY_COGNITO_PROVIDER =
+        new CognitoUserPoolsAuthProvider() {
+            @Override
+            public String getLatestAuthToken() throws ApiException {
+                return null;
+            }
+
+            @Override
+            public String getUsername() {
+                return null;
+            }
+        };
+
+    /**
+     * Test setup.
+     * @throws AmplifyException not expected.
+     */
+    @Before
+    public void setup() throws AmplifyException {
+        Amplify.removePlugin(AUTH_PLUGIN);
+    }
+
+    /**
+     * Verifies that API_KEY is returned as one of the available auth types if:
+     * - An API key auth provider is explicitly set on the ApiAuthProviders instance OR
+     * - The provided API configuration has an API key specified.
+     */
+    @Test
+    public void testApiKey() {
+        // Create an API config with an API key
+        ApiConfiguration configWithKey = ApiConfiguration.builder()
+                                                         .endpoint("https://test.endpoint.com")
+                                                         .region("us-east-1")
+                                                         .authorizationType(AuthorizationType.API_KEY)
+                                                         .apiKey("dummy")
+                                                         .build();
+
+        // Create an ApiAuthProviders instance with an API key provider set.
+        ApiAuthProviders withProvider = ApiAuthProviders.builder().apiKeyAuthProvider(() -> null).build();
+
+        // Create an empty ApiAuthProviders instance.
+        ApiAuthProviders withoutProvider = ApiAuthProviders.noProviderOverrides();
+
+        List<AuthorizationType> expectedAuthTypesApiKeyOnly = Collections.singletonList(AuthorizationType.API_KEY);
+
+        // No API Key provider and no API key in the config should result in 0 auth types being returned.
+        List<AuthorizationType> actualAuthTypes = withoutProvider.getAvailableAuthorizationTypes(BASIC_API_CONFIG);
+        assertEquals(0, actualAuthTypes.size());
+
+        // No API Key provider but there's an API key in the config should result in 1 auth type (API_KEY)
```

Comment:
Can you explain this behavior? 
Hypothetically, if a user 
1. starts out with API key mode for testing, 
2. manually switches to a more secure auth mode such as user pools, 
3. and then forgets to remove the key from config, 

then are we treating it as if the customer set up a multi-auth? Shouldn't the customer be required to be explicit regarding which modes of auth types to be available for the plugin? In above scenario, wouldn't it be possible for the customer to not realize that API key mode is being used unintentionally by the plugin for certain operations that should not have been authorized?

Reply:
It's possible. However, this covers the use case that's probably more common where the API key is in the config but the customer does not set the apiKeyProvider in the ApiAuthProviders passed in to the constructor (or the builder now). I think that in most of our docs, the examples don't show the API plugin getting initialized with an explicitly provided ApiAuthProviders object.

Also, having multiple providers available doesn't necessarily mean they're using multi-auth. In non-multiauth scenarios, each request will only be attempted with one auth type, but that auth type can be any of the ones configured in ApiAuthProviders. 

For example, if the API is setup with user pools as default and API key as additional, any API requests where no auth type is specified will use the default. If an API call is made where the caller specifies the auth type (API key for instance), that request will use that auth type if the provider is configured.